From a1ec170d10b2989c238bb58a8aad754f9a3516b3 Mon Sep 17 00:00:00 2001
From: Jens Reidel <adrian@travitia.xyz>
Date: Wed, 13 Aug 2025 20:30:48 +0200
Subject: [PATCH] hexagonrpcd: Guess directory for hexagonfs from compatible

Signed-off-by: Jens Reidel <adrian@travitia.xyz>
---
 hexagonrpcd/rpcd.c | 104 +++++++++++++++++++++++++++++++++++++++++++++
 1 file changed, 104 insertions(+)

diff --git a/hexagonrpcd/rpcd.c b/hexagonrpcd/rpcd.c
index 05a8611..4293f0f 100644
--- a/hexagonrpcd/rpcd.c
+++ b/hexagonrpcd/rpcd.c
@@ -292,12 +292,112 @@ err:
 	return NULL;
 }
 
+static char *read_sysfs_file(const char *path, struct stat *file_stat)
+{
+	char *contents = NULL;
+	int file_fd;
+
+	if ((file_fd = open(path, O_RDONLY)) == -1)
+		return NULL;
+
+	if (fstat(file_fd, file_stat))
+		goto close_fd;
+
+	contents = malloc(file_stat->st_size);
+	if (contents == NULL)
+		goto close_fd;
+
+	if (read(file_fd, contents, file_stat->st_size) != file_stat->st_size) {
+		free(contents);
+		contents = NULL;
+	}
+
+close_fd:
+	close(file_fd);
+
+	return contents;
+}
+
+static char *guess_device_directory_from_compatible(void)
+{
+	char *compatible_ptr, *device_name, *model_ptr, *orig_compatible_ptr,
+		*end_of_vendor_name, *vendor_name;
+	char *chipset_name = NULL;
+	char *ret = NULL;
+	struct stat compatible_stat, model_stat;
+
+	if ((compatible_ptr = read_sysfs_file("/proc/device-tree/compatible",
+					      &compatible_stat)) == NULL)
+		return NULL;
+	orig_compatible_ptr = compatible_ptr;
+
+	if ((model_ptr = read_sysfs_file("/proc/device-tree/model",
+					 &model_stat)) == NULL)
+		goto free_compatible;
+
+	// Get capitalized vendor name from model string
+	end_of_vendor_name = strchr(model_ptr, ' ');
+	if (end_of_vendor_name == NULL)
+		end_of_vendor_name = model_ptr + model_stat.st_size - 1;
+	vendor_name = strndup(model_ptr, end_of_vendor_name - model_ptr);
+	if (vendor_name == NULL)
+		goto free_model;
+
+	// Get the chipset name (usually after the model name)
+	while (compatible_ptr < orig_compatible_ptr + compatible_stat.st_size) {
+		if (strncmp("qcom,", compatible_ptr, sizeof("qcom,") - 1) ==
+		    0) {
+			chipset_name = compatible_ptr + sizeof("qcom,") - 1;
+			break;
+		}
+
+		compatible_ptr += strlen(compatible_ptr) + 1;
+	}
+
+	if (chipset_name == NULL)
+		goto free_model;
+
+	// Try all other compatible entries as device codenames
+	compatible_ptr = orig_compatible_ptr;
+	while (compatible_ptr < orig_compatible_ptr + compatible_stat.st_size) {
+		char *try_path = NULL;
+
+		// Split the entry at the ,
+		device_name = strchr(compatible_ptr, ',');
+		if (device_name == NULL)
+			continue;
+		device_name += 1; // Skip the ,
+
+		// Check if this soc + vendor + device combo path exists
+		if (asprintf(&try_path, "/usr/lib/firmware/hexagonfs/%s/%s/%s",
+			     chipset_name, vendor_name, device_name) < 0)
+			goto free_model;
+
+		if (!access(try_path, R_OK)) {
+			ret = try_path;
+			break;
+		}
+
+		free(try_path);
+
+		compatible_ptr += strlen(compatible_ptr) + 1;
+	}
+
+free_model:
+	free(model_ptr);
+free_compatible:
+	free(orig_compatible_ptr);
+
+	return ret;
+}
+
 int main(int argc, char* argv[])
 {
 	char *fastrpc_node = NULL;
 	const char *device_dir = "/usr/share/qcom/";
 	const char *dsp = "";
 	const char *create_shell = NULL;
+	const char *guessed_device_dir;
 	const char **progs;
 	pid_t *pids;
 	size_t n_progs = 0;
@@ -316,6 +416,10 @@ int main(int argc, char* argv[])
 		goto err_free_progs;
 	}
 
+	guessed_device_dir = guess_device_directory_from_compatible();
+	if (guessed_device_dir != NULL)
+		device_dir = guessed_device_dir;
+
 	while ((opt = getopt(argc, argv, "c:d:f:p:R:s")) != -1) {
 		switch (opt) {
 			case 'c':
-- 
2.50.1

